src/io.c

```diff
@@ -196,10 +241,60 @@ Obj FuncIO_RestoreSIGCHLDHandler( Obj self )
   }
 }
 
+// The following function checks if a PID is marked as ignored.
+// Returns 1 if the PID was ignored, 0 otherwise.
+static int IO_CheckForIgnoredPid( int pid )
+{
+    int i;
+    // Make sure a new signal doesn't come in while looking at array
+    signal(SIGCHLD, SIG_DFL);
+    int found = 0;
+    for(i = 0; i < ignoredpidslen; ++i) {
+        if (ignoredpids[i] == pid) {
+            ignoredpids[i] = ignoredpids[ignoredpidslen - 1];
+            ignoredpidslen--;
+            found = 1;
+            break;
+        }
+    }
+    signal(SIGCHLD,IO_SIGCHLDHandler);
+    return found;
+}
+
+Obj FuncIO_IgnorePid(Obj self, Obj pid)
+{
+    Int pidc;
+    int pos;
+    if (!IS_INTOBJ(pid)) {
+        return Fail;
+    }
+    pidc = INT_INTOBJ(pid);
+
+    pos = findSignaledPid(pidc);
+    if(pos != -1)
+    {
+        // This PID has already finished
+        removeSignaledPidByPos(pos);
+        return True;
+    }
+    if (ignoredpidslen < MAXCHLDS - 1) {
+        // Make sure a new signal doesn't come in while we are changing array
+        signal(SIGCHLD, SIG_DFL);
+        ignoredpids[ignoredpidslen] = pidc;
+        ignoredpidslen++;
+        signal(SIGCHLD,IO_SIGCHLDHandler);
```

Comment:
Yeah, I think those signal handler calls need to be moved to the start resp. end of the function.